ocaml/xapi/xapi_cluster.ml

```diff
@@ -108,103 +112,82 @@ let pool_create ~__context ~network ~cluster_stack ~token_timeout ~token_timeout
   in
 
   List.iter (fun host ->
-      (* Cluster.create already created cluster_host on master, so we only need to iterate through slaves *)
+      (* Cluster.create already created cluster_host on master, so we only iterate through slaves *)
       Helpers.call_api_functions ~__context (fun rpc session_id ->
-          let pifref,_ = pif_of_host ~__context network host in
-          let cluster_host_ref = Client.Client.Cluster_host.create ~rpc ~session_id ~cluster ~host ~pif:pifref in
+          let pif,_ = pif_of_host ~__context network host in
+          let cluster_host_ref = Client.Client.Cluster_host.create ~rpc ~session_id ~cluster ~host ~pif in
           D.debug "Created Cluster_host: %s" (Ref.string_of cluster_host_ref);
         )) slave_hosts;
 
   cluster
 
-(* Helper function; if opn is None return all, else return those not equal to it *)
-let filter_on_option opn xs =
-  match opn with
-  | None -> xs
-  | Some x -> List.filter ((<>) x) xs
+let foreach_cluster_host ~__context ~self ~(fn : rpc:(Rpc.call -> Rpc.response) ->
+      session_id:API.ref_session -> self:API.ref_Cluster_host -> unit) ~log =
+  let wrapper = if log then log_and_ignore_exn else (fun _ -> ()) in
```

Comment:
The bug is here: if log is false then this does not run `fn` at all, the correct way would be:
`let wrapper = if log then log_and_ignore_exn else (fun f -> f ()) in`

Reply:
time to hang my head in shame
or borrow the cone of shame from my dog